feature(security): Adds functions to create and validate HMAC tokens

mrclay opened the pull request:
This also has getHmac() use a binary encoding of the site key instead of the Base64 or hex encodings.

Fixes #7824

jdalsem commented:
Does not really fix #7824 as site secret is still not publicly through API, but this PR will reduce the need to have it available a bit. Don't know if it will cover all usecases.

ewinslow commented on engine/tests/phpunit/ElggCryptoTest.php:
```diff
@@ -41,4 +45,31 @@ function provider() {
+		$this->assertTrue((bool)preg_match('~^[a-zA-Z0-9\-_]{30,}$~', $mac));
+
+		$mac = $crypto->getHmac([1234, 'foo'], $key);
+		$this->assertTrue((bool)preg_match('~^[a-zA-Z0-9\-_]{30,}$~', $mac));
```
What is this magic regex?

ewinslow commented:
Looks cool. Bummer about the new global method but whatevs I guess.

jdalsem commented on engine/classes/Elgg/ActionsService.php:
```diff
@@ -264,10 +264,10 @@ public function generateActionToken($timestamp) {
 		$site_secret = _elgg_services()->siteSecret->get();
```
i believe you can drop this, as you are now defaulting

jdalsem commented on engine/lib/actions.php:
```diff
@@ -98,6 +98,38 @@ function elgg_unregister_action($action) {
+	return $crypto->areEqual($mac, $crypto->getHmac($data, '', $algo));
+}
+
+/**
```
do these functions really need to be in actions? I don't know a really good location either

jdalsem commented:
> Don't know if it will cover all usecases.

i check with all our own usecases and it covers them all. So really happy to see this in core

juho-jaakkola commented on docs/guides/actions.rst:
```diff
@@ -425,3 +425,30 @@ You can also access the tokens from javascript:
 
+Security Tokens
+===============
+If you need to pass data to a user and trust that the user has returned it to you unaltered, you should use
```
I had to read this twice to understand it. Could this be explained better?

mrclay committed:
feature(security): Adds component to create and validate HMAC tokens

HMAC uses a binary encoding of the site key instead of the Base64 or hex encodings, and improves the docs of the site secret component.

Fixes #7824

mrclay commented:
I've tried to address all these. Note that this will invalidate any unused tokens such as email validations/invitations. It uses a binary form of the site key instead of the hex/b64 encodings and also includes separators in the data.

I walked back the idea of auto-serializing data passed in because I was uncomfortable with having HMAC not match other implementations, but I'm still a bit torn on this. If someone fails to use a good separator, an attacker can move characters between variables: E.g. using `$data = $a . $b` means if you have a MAC for "123" and "hello", you can re-use that MAC for "12" and "3hello". That's not good.

@ewinslow Re global functions: https://community.elgg.org/discussion/view/2062019/ideas-for-exposing-elgg-oo-services

beck24 commented:
LGTM

juho-jaakkola commented on engine/classes/Elgg/Database/SiteSecret.php:
```diff
@@ -36,14 +36,32 @@ function init() {
+		if ($raw) {
+			// try to return binary key
+			if ($secret[0] === 'z') {
+				// new keys are "z" + base64URL
```
Why z? Might be best to document.

mrclay commented:
I'm thinking of an API like:

```php
// HMAC with direct set of data
$token = elgg_build_hmac()->setData("$foo|$bar")->getToken();

$is_valid = elgg_build_hmac()->setData("$foo|$bar")->matchesToken($token);

// preferred usage: handles separation for you
$token = elgg_build_hmac()
    ->addData($foo)
    ->addData($bar)
    ->getToken();
```

mrclay commented:
I reworked this so that it can accept an array of strings as HMAC data, and spits out an object instead of breaking generation/validation into two functions.

juho-jaakkola commented:
LGTM

hypeJunction commented:
I am unable to get this to work. For whatever reasons tokens do not match. Using hash_hmac() for now.

mrclay commented:
@hypeJunction did you make sure the types match? http://learn.elgg.org/en/1.11/guides/actions.html#security-tokens

hypeJunction commented:
I did.
On Apr 15, 2015 4:43 AM, "Steve Clay" <notifications@github.com> wrote:

> @hypeJunction <https://github.com/hypeJunction> did you make sure the
> types match?
> http://learn.elgg.org/en/1.11/guides/actions.html#security-tokens
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8057#issuecomment-93162872>.
>

hypeJunction commented:
I am not sure if this affects the matching alrgorithm, but link was generated with elgg_build_hmac() and validation was being done without engine boot using Hmac component.

mrclay commented:
Ah, few things: server key is decoded from b64/hex to binary before use, and the hash created is b64URL encoded. I should've made all those steps easily accessible without boot.

mrclay commented:
At the very least I'll add that to docs for this function
